medusa/themes/base.py

```diff
@@ -0,0 +1,73 @@
+import logging
+import json
+import os
+from medusa import app
+from medusa.logger.adapters.style import BraceAdapter
+
+
+log = BraceAdapter(logging.getLogger(__name__))
+log.logger.addHandler(logging.NullHandler())
+
+themes = []
+
+
+class Theme(object):
+    """Base theme class."""
+
+    def __init__(self, name, version, dep_version=None, author=None):
+        self.name = name
+        self.version = version
+        self.dep_version = dep_version
+
+
+def read_themes():
+    """Read the theme folder and create a Theme object for every entry, if valid."""
+    theme_root = os.path.join(app.PROG_DIR, 'themes')
+    themes_from_fs = [os.path.join(theme_root, theme_dir) for theme_dir in os.listdir(theme_root) if os.path.isdir(os.path.join(theme_root, theme_dir))]
+    for theme_path in themes_from_fs:
+        # validate the directory structure
+        try:
+            validate_theme(theme_path)
+        except Exception as error:
+            log.error('Error reading theme {theme}, {error!r}', {'theme': theme_path, 'error': error})
+            raise Exception('Error in one of the theme paths. Please fix the error and start medusa.')
+
+        # validate the manifest
+        manifest = json.load(open(os.path.join(theme_path, 'manifest.json')))
+        if theme_path:
+            name = manifest['name']
+            version = manifest['version']
+            themes.append(Theme(name, version))
+
+    return themes
+
+
+def validate_theme(theme_path):
+    """Validate theme configuration."""
+    if not os.path.isdir(os.path.join(theme_path, 'assets')):
+        raise Exception('Missing assets folder. Please refer to the medusa theming documentation.')
+
+    for check_folder in ('css', 'js', 'img'):
+        if not os.path.isdir(os.path.join(theme_path, 'assets', check_folder)):
+            raise Exception('Missing folder: [theme]/assets/{check_folder}. '
+                            'Please refer to the medusa theming documentation.'.format(check_folder=check_folder))
+
+    try:
+        manifest = json.load(open(os.path.join(theme_path, 'manifest.json')))
+    except IOError:
+        raise Exception('Cannot read manifest.json. Please refer to the medusa theming documentation.')
+
+    # Validate if they mandatory keys are configured in the manifest.
+    if not manifest.get('name') or not manifest.get('version'):
+        raise Exception("As a bare minimum you'l need at least to provide the 'name' and and 'version' key. "
+                        "Please refer to the medusa theming documentation.")
+
+    # The theme's folder name should be the same as the name in the manifest
+    if os.path.basename(os.path.normpath(theme_path)) != manifest['name']:
```

Comment:
Is there actually a need for this check? It'd mean the name can't contain special characters for example that Windows won't allow.

Reply:
Consistency. As both the folder name and the manifest name are used.